pkg/config/tlscfg/reload.go

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) 2020 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tlscfg
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io"
+	"path/filepath"
+	"sync"
+
+	"github.com/fsnotify/fsnotify"
+	"go.uber.org/zap"
+)
+
+type watchCerts struct {
+	opts    Options
+	watcher *fsnotify.Watcher
+	cert    tls.Certificate
+	logger  *zap.Logger
+	mu      *sync.Mutex
+}
+
+var _ io.Closer = (*watchCerts)(nil)
+
+func newWatchCerts(opts Options, logger *zap.Logger) (*watchCerts, error) {
+	var cert tls.Certificate
+	if opts.CertPath != "" && opts.KeyPath != "" {
+		// load certs at startup to catch missing certs error early
+		c, err := tls.LoadX509KeyPair(filepath.Clean(opts.CertPath), filepath.Clean(opts.KeyPath))
+		if err != nil {
+			return nil, fmt.Errorf("failed to load server TLS cert and key: %w", err)
+		}
+		cert = c
+	}
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+	if err := addCertsToWatch(watcher, opts); err != nil {
+		return nil, err
+	}
+	return &watchCerts{
+		cert:    cert,
+		opts:    opts,
+		watcher: watcher,
+		logger:  logger,
+		mu:      &sync.Mutex{},
+	}, nil
+}
+
+func (w watchCerts) Close() error {
+	return w.watcher.Close()
+}
+
+func (w watchCerts) certificate() *tls.Certificate {
+	w.mu.Lock()
```

Comment:
Did you consider using a [`RWMutex`](https://golang.org/pkg/sync/#RWMutex) instead? You can just lock it for read here, so that you can have concurrent reads.